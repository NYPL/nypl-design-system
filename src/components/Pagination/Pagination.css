@import "../../styles/base/03-mixins";

.pagination {
  align-items: stretch;
  display: flex;
  margin-bottom: var(--space-l);
  width: 100%;

  > * {
    &:not(:last-child) {
      @include space-inline-xs;
    }
  }

  &__list {
    @include list-reset;

    display: flex;
  }

  &__item {
    list-style-type: none;
  }

  &__link {
    line-height: 1.15;
    padding: 8px;
    text-decoration: none;

    &.selected {
      color: var(--ui-black);
    }

    &.disabled {
      color: var(--ui-gray-dark);
      pointer-events: none;
    }
  }
}
