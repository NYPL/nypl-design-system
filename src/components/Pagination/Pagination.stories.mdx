import {
  Meta,
  Story,
  ArgsTable,
  Preview,
  Description,
} from "@storybook/addon-docs/blocks";
import { withDesign } from "storybook-addon-designs";
import * as stories from "./Pagination.stories.tsx";

<Meta
  title="Pagination"
  component={Pagination}
  decorators={[withDesign]}
  parameters={{
    design: {
      type: "figma",
      url:
        "https://www.figma.com/file/qShodlfNCJHb8n03IFyApM/Master?node-id=17226%3A932",
    },
  }}
/>

# Pagination

<Description of={Pagination} />
Pagination navigates between the pages of a multi-page application. It is commonly
in search results. Use controls in the Basic example to explore this component with
very many or very few pages.

Pagination uses anchor tags because it is navigating between URLs.

<Preview>
  <Story story={stories.paginationUrlHref} />
</Preview>

## Pagination with onClick

It is preferred to use pagination with `getPageHref` and that the URL is updated whenever a page is changed.
However, optional `onClick` is available in case URL updating is not desired.

<Preview>
  <Story story={stories.paginationOnClick} />
</Preview>
