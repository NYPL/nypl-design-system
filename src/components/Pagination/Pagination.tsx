import * as React from "react";
import range from "lodash/range";
import Link from "../Link/Link";
import bem from "../../utils/bem";

export interface PaginationProps {
  /** BlockName for use with BEM. See how to work with blockNames and BEM here: http://getbem.com/introduction/ */
  blockName?: string;
  /** Additional className for use with BEM. See how to work with blockNames and BEM here: http://getbem.com/introduction/ */
  className?: string;
  /** Modifiers array for use with BEM. See how to work with modifiers and BEM here: http://getbem.com/introduction/ */
  modifiers?: string[];
  /** The total number of pages. */
  pageCount: number;
  /** The current page selected. */
  currentPage: number;
  /** A method that returns a Link component given the target page */
  getPageHref?: (pageNumber: number) => string;
  /** The method to callback when an item is selected. Passes the selected page to the consuming app as an argument. */
  onPageChange?: (selected: number) => void;
}

/** Pagination component that provides a list of page items */
const Pagination: React.FC<PaginationProps> = (props: PaginationProps) => {
  const {
    blockName,
    className,
    modifiers = [],
    pageCount,
    currentPage = 1,
    getPageHref,
    onPageChange,
  } = props;

  //If there are 0 or 1 pages, the pagination should not show
  if (pageCount <= 1) {
    return null;
  }

  if (getPageHref && onPageChange) {
    console.warn(
      "Both onPageHref and onPageChange are passed. Will default to using onPageHref"
    );
  }

  const changeUrls =
    typeof getPageHref !== "undefined" && typeof getPageHref === "function";

  const selectPage = (evt: Event, item: number) => {
    const runCallback = (selectedItem: number) => {
      evt.preventDefault ? evt.preventDefault() : (evt.returnValue = false);

      if (
        typeof onPageChange !== "undefined" &&
        typeof onPageChange === "function"
      ) {
        onPageChange(selectedItem);
      }
    };

    evt.preventDefault ? evt.preventDefault() : (evt.returnValue = false);
    if (currentPage === item) return;

    // Run the callback with the new selected item:
    runCallback(item);
  };

  const previousPage = (evt: Event) => {
    evt.preventDefault ? evt.preventDefault() : (evt.returnValue = false);
    if (currentPage > 1) {
      selectPage(evt, currentPage - 1);
    }
  };

  const nextPage = (evt: Event) => {
    evt.preventDefault ? evt.preventDefault() : (evt.returnValue = false);
    if (currentPage < pageCount) {
      selectPage(evt, currentPage + 1);
    }
  };

  const getPageElement = (item: number) => {
    const pageAttributes = {
      "aria-label": null,
      onClick: !changeUrls ? evt => selectPage(evt, item) : undefined,
      role: "button",
      tabIndex: 0,
    };

    pageAttributes["aria-label"] = item ? item : null;

    const pageClass = currentPage === item ? "selected" : null;

    return (
      <Link
        attributes={{ ...pageAttributes }}
        className={bem("link", modifiers, "pagination", [pageClass])}
        href={changeUrls ? getPageHref(item) : "#"}
      >
        {item}
      </Link>
    );
  };

  const pagination = (selected: number) => {
    // 1, (2 or ...), pageStart - pageEnd, (next to last item or ...), pageCount
    const pageStart = Math.max(
      3, // if near the beginning, inner pages start at 3
      Math.min(
        // If in the middle, inner pages begin two items before the selected page
        selected - 2,
        // If near the end, count back and show 7 items
        pageCount - 6
      )
    );

    const pageEnd = Math.min(
      // if near the end, inner pages end just before next to last item (or ellipse)
      pageCount - 2,
      Math.max(
        // If in the middle, inner pages end two items after the selected page
        selected + 2,
        // If near the beginning, show the first 7 items
        7
      )
    );

    const truncatedList =
      pageCount < 4
        ? Array.from({ length: pageCount }, (_, i) => i + 1)
        : [
            // list always starts at and displays the first page item
            1,
            // second item will either be item 2 or an ellipse
            pageStart > 3 ? "ellipse-start" : 2,
            // Adding + 1 here since lodash range() doesn't include the last item passed to it
            ...range(pageStart, pageEnd + 1),
            // next to last item will either be an ellipse or the next to last page number
            pageEnd < pageCount - 2 ? "ellipse-end" : pageCount - 1,
            // list always ends at and displays the last page number in the pageCount
            pageCount,
          ];
    const pageItems = truncatedList.map(item => {
      // if it's a number, render that page item, otherwise return the ellipse
      const itemElement =
        typeof item === "number" ? getPageElement(item) : "...";
      return (
        <li key={item} className={bem("item", modifiers, "pagination")}>
          {itemElement}
        </li>
      );
    });

    return pageItems;
  };

  // When at the beginning, disable Previous. When at the end, disable Next.
  const prevDisabled = currentPage === 1;
  const nextDisabled = currentPage === pageCount;

  // Attributes for Previous and Next Buttons
  const prevAttributes = {
    "aria-disabled": prevDisabled ? "true" : null,
    "aria-label": "Previous page",
    role: "button",
    onClick: changeUrls ? undefined : previousPage,
    tabIndex: prevDisabled ? -1 : 0,
  };

  const nextAttributes = {
    "aria-disabled": nextDisabled ? "true" : null,
    "aria-label": "Next page",
    role: "button",
    onClick: changeUrls ? undefined : nextPage,
    tabIndex: nextDisabled ? -1 : null,
  };

  const prevClass = prevDisabled ? "disabled" : null;
  const nextClass = nextDisabled ? "disabled" : null;

  return (
    <nav
      aria-label="Pagination"
      className={bem("pagination", modifiers, blockName, [className])}
    >
      <ul className={bem("list", modifiers, "pagination")}>
        <li key="previous" className={bem("item", modifiers, "pagination")}>
          <Link
            attributes={{ ...prevAttributes }}
            className={bem("link", modifiers, "pagination", [prevClass])}
            href={
              !prevDisabled && changeUrls ? getPageHref(currentPage - 1) : "#"
            }
          >
            Previous
          </Link>
        </li>

        {pagination(currentPage)}

        <li key="next" className={bem("item", modifiers, "pagination")}>
          <Link
            attributes={{
              ...nextAttributes,
            }}
            className={bem("link", modifiers, "pagination", [nextClass])}
            href={
              !nextDisabled && changeUrls ? getPageHref(currentPage + 1) : "#"
            }
          >
            Next
          </Link>
        </li>
      </ul>
    </nav>
  );
};

export default Pagination;
