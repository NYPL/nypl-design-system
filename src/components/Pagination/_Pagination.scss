.pagination {
  @include space-stack-l;

  align-items: stretch;
  display: flex;
  width: 100%;

  > * {
    &:not(:last-child) {
      @include space-inline-xs;
    }
  }

  &__list {
    @include list-reset;

    display: flex;
  }

  &__item {
    @include space-inline-xs;

    list-style-type: none;
  }

  &__link {
    line-height: 1.15;
    text-decoration: none;

    &.selected {
      color: var(--ui-black);
      pointer-events: none;
    }

    &.disabled {
      color: var(--ui-gray-dark);
      pointer-events: none;
    }
  }
}
