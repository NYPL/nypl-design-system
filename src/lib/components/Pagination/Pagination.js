"use strict";
var __createBinding =
    (this && this.__createBinding) ||
    (Object.create
        ? function (o, m, k, k2) {
              if (k2 === undefined) k2 = k;
              Object.defineProperty(o, k2, {
                  enumerable: true,
                  get: function () {
                      return m[k];
                  },
              });
          }
        : function (o, m, k, k2) {
              if (k2 === undefined) k2 = k;
              o[k2] = m[k];
          });
var __setModuleDefault =
    (this && this.__setModuleDefault) ||
    (Object.create
        ? function (o, v) {
              Object.defineProperty(o, "default", {
                  enumerable: true,
                  value: v,
              });
          }
        : function (o, v) {
              o["default"] = v;
          });
var __importStar =
    (this && this.__importStar) ||
    function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (Object.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    };
var __importDefault =
    (this && this.__importDefault) ||
    function (mod) {
        return mod && mod.__esModule ? mod : { default: mod };
    };
Object.defineProperty(exports, "__esModule", { value: true });
const React = __importStar(require("react"));
const bem_1 = __importDefault(require("../../utils/bem"));
const Button_1 = __importDefault(require("../Button/Button"));
const Dropdown_1 = __importDefault(require("../Dropdown/Dropdown"));
const ButtonTypes_1 = require("../Button/ButtonTypes");
const IconTypes_1 = require("../Icons/IconTypes");
/** Pagination */
function Pagination(props) {
    const {
        paginationDropdownOptions,
        currentValue,
        previousPageHandler,
        nextPageHandler,
        onSelectChange,
        onSelectBlur,
    } = props;
    const pagination__base_class = "pagination";
    let buttonPrevOpts = {
        id: "pagination-previous",
        callback: previousPageHandler,
        content: React.createElement(React.Fragment, null, "Previous"),
        buttonType: ButtonTypes_1.ButtonTypes.Outline,
        iconPosition: ButtonTypes_1.ButtonIconPositions.Left,
        iconRotation: IconTypes_1.IconRotationTypes.rotate90,
        iconName: "arrow_xsmall",
        iconDecorative: true,
    };
    let buttonNextOpts = {
        id: "pagination-next",
        callback: nextPageHandler,
        content: React.createElement(React.Fragment, null, "Next"),
        buttonType: ButtonTypes_1.ButtonTypes.Outline,
        iconPosition: ButtonTypes_1.ButtonIconPositions.Right,
        iconRotation: IconTypes_1.IconRotationTypes.rotate270,
        iconName: "arrow_xsmall",
        iconDecorative: true,
    };
    let dropdownProps = {
        dropdownId: "pagination-dropdown",
        isRequired: false,
        labelId: "pagination-select-label",
        labelText: "Page",
        dropdownOptions: paginationDropdownOptions,
        selectedOption: currentValue,
        onSelectChange: onSelectChange,
        onSelectBlur: onSelectBlur,
    };
    return React.createElement(
        "nav",
        {
            "aria-label": "Pagination",
            className: bem_1.default(pagination__base_class),
        },
        React.createElement(
            Button_1.default,
            Object.assign(
                { callback: buttonPrevOpts.callback },
                buttonPrevOpts
            ),
            buttonPrevOpts.content
        ),
        React.createElement(
            Dropdown_1.default,
            Object.assign(
                { blockName: pagination__base_class, labelPosition: "left" },
                dropdownProps
            )
        ),
        React.createElement(
            Button_1.default,
            Object.assign(
                { callback: buttonNextOpts.callback },
                buttonNextOpts
            ),
            buttonNextOpts.content
        )
    );
}
exports.default = Pagination;
