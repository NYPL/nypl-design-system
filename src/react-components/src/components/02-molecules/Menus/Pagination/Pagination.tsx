import * as React from "react";
import bem from "../../../../utils/bem";
import Button, { ButtonOptions } from "../../../01-atoms/Button/Button";
import Dropdown from "../../../02-molecules/Dropdown/Dropdown";
import {
    ButtonIconPositions,
    ButtonTypes,
} from "../../../01-atoms/Button/ButtonTypes";

export interface PaginationProps {
    paginationDropdownOptions: string[];
    previousPageHandler: (event: React.MouseEvent) => void;
    nextPageHandler: (event: React.MouseEvent) => void;
    currentValue: string;
    onSelectChange: (event: React.MouseEvent) => void;
    onSelectBlur: (event: React.MouseEvent) => void;
}

/** Pagination */
export default function Pagination(props: PaginationProps) {
    const {
        paginationDropdownOptions,
        currentValue,
        previousPageHandler,
        nextPageHandler,
        onSelectChange,
        onSelectBlur,
    } = props;

    const pagination__base_class = "pagination";
    let buttonPrevOpts: ButtonOptions = {
        id: "pagination-previous",
        callback: previousPageHandler,
        content: <>Previous</>,
        buttonType: ButtonTypes.Outline,
        iconPosition: ButtonIconPositions.Left,
        iconName: "arrow_xsmall",
        iconModifiers: ["left", "icon-left"],
        iconDecorative: true,
    };

    let buttonNextOpts: ButtonOptions = {
        id: "pagination-next",
        callback: nextPageHandler,
        content: <>Next</>,
        buttonType: ButtonTypes.Outline,
        iconPosition: ButtonIconPositions.Right,
        iconName: "arrow_xsmall",
        iconModifiers: ["right", "icon-right"],
        iconDecorative: true,
    };

    let dropdownProps = {
        dropdownId: "pagination-dropdown",
        isRequired: false,
        labelId: "pagination-select-label",
        labelText: "Page",
        dropdownOptions: paginationDropdownOptions,
        selectedOption: currentValue,
        onSelectChange: onSelectChange,
        onSelectBlur: onSelectBlur,
    };

    return (
        <nav aria-label="Pagination" className={bem(pagination__base_class)}>
            <Button callback={buttonPrevOpts.callback} {...buttonPrevOpts}>
                {buttonPrevOpts.content}
            </Button>
            <Dropdown
                blockName={pagination__base_class}
                labelPosition="left"
                {...dropdownProps}
            ></Dropdown>
            <Button callback={buttonNextOpts.callback} {...buttonNextOpts}>
                {buttonNextOpts.content}
            </Button>
        </nav>
    );
}
