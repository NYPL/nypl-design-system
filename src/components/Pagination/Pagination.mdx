import { ArgTypes, Canvas, Description, Meta, Source } from "@storybook/blocks";

import * as PaginationStories from "./Pagination.stories";

<Meta of={PaginationStories} />

# Pagination

| Component Version | DS Version |
| ----------------- | ---------- |
| Added             | `0.0.10`   |
| Latest            | `1.5.0`    |

## Table of Contents

- [Overview](#overview)
- [Accessibility](#accessibility)
- [Pagination with URL Updates](#pagination-with-url-updates)
- [Pagination with Unchanging URL](#pagination-with-unchanging-url)

## Overview

<Description of={PaginationStories} />

The `Pagination` component helps navigate between pages of a multi-page
application. It is commonly used on a search results page. Update the `pageCount`
prop in the Controls to explore this component with many or few pages.

## Accessibility

Internally, the `Pagination` component is implemented with a `<nav>` element with
an `aria-label` attribute of `"Pagination"` and an unordered `<ul>` element. This
component uses anchor `<a>` tags because it is navigating between URLs. In the
"unchanging URL" variation, each anchor tag has an `href` attribute with a value
of `"#"`, because the URL is not changing.

Resources:

- [W3C Design System Pagination](https://design-system.w3.org/components/pagination.html)
- [a11ymatters Accessible Pagination](https://www.a11ymatters.com/pattern/pagination/)

## Pagination with URL Updates

In order to update the URL with a query parameter, the parent component must
declare and pass a `getPageHref` function to the `Pagination` component. This
function takes a `selectedPage` number argument for the selected page and it
should return a string with the desired URL that the pagination links should go
to. Note that this will refresh the browser on every click.

In the following example, the `getPageHref` function will compute a new URL that
has `&page=` as a URL query. In your own application this query parameter can be
named anything but you have to pick up the value in the wrapper component.

<Source
  code={`
// Example in a search results page.
const getPageHref = (selectedPage: number) => {
  // This should be updated for your router system, if any, including the
  // base or origin URL.
  return \`{location.origin}?q=celeste&page=\$\{selectedPage\}\`;
};
`}
  language="jsx"
/>

<Canvas of={PaginationStories.URLUpdates} />

<ArgTypes of={PaginationStories.URLUpdates} />

## Pagination with Unchanging URL

It is preferred to use the `Pagination` component with the `getPageHref`
function passed as a prop and that the URL is updated whenever a page is
changed. However, the optional `onPageChange` prop is available in case URL
updating is not desired.

In the following example, the `onPageChange` function gets the selected page as
its only function argument. This is computed internally in the `Pagination`
component through its own state.

<Source
  code={`
// Example in a search results page.
const onPageChange = (selectedPage: number) => {
  console.log(\`Current page: \$\{selectedPage\}\`);
  // Do what you need to with the \`currentPage\` value.
};
`}
  language="jsx"
/>

<Canvas of={PaginationStories.UnchangingURL} />

### Forcing a Page Change Programmatically

There may be circumstances when you want to programmatically force the page
number to change without the user explicitly requesting it (for example, if
you want a user to be brought back to page 1 after entering a new search term).
You can use the `currentPage` prop to do this. Note, the `currentPage` prop can
only be used with the client-side, unchanging URL version of the `Pagination`
component, and in combination with the `onPageChange` function.

The following example shows that the user can be brought to page 1 without
clicking on the `Pagination` component.

<Canvas of={PaginationStories.UpdateProgrammatically} />
