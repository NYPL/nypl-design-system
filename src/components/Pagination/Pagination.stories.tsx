import React, { useEffect, useState } from "react";
import { Story } from "@storybook/react/types-6-0";
import { withDesign } from "storybook-addon-designs";

import Pagination, { PaginationProps } from "./Pagination";
import { withQuery } from "@storybook/addon-queryparams";

export default {
  title: "Pagination",
  component: Pagination,
  decorators: [withDesign, withQuery],
};

const getPageNumber = (page: number, pageCount: number) => {
  return page > 0 && page < pageCount ? page : 1;
};

const PaginationPageHrefTemplate: Story<PaginationProps> = args => {
  const urlParams = new URLSearchParams(document.location.search);
  const pageParam = urlParams.get("page");
  const pageNumber =
    pageParam &&
    Number(pageParam) &&
    getPageNumber(Number(pageParam), args.pageCount);

  const location = window.location;

  console.log("pageNumber", pageNumber);
  const getPageHref = page => {
    const currentStoryId = urlParams.get("id");
    return `${location.origin}?path=/story/${currentStoryId}&page=${page}`;
  };
  return (
    <Pagination
      pageCount={args.pageCount}
      currentPage={pageNumber}
      getPageHref={getPageHref}
    />
  );
};

export const paginationUrlHref = PaginationPageHrefTemplate.bind({});

paginationUrlHref.args = {
  currentPage: 4,
  pageCount: 100,
};

paginationUrlHref.argTypes = {
  blockName: { table: { disable: true } },
  className: { table: { disable: true } },
  modifiers: { table: { disable: true } },
  currentPage: { table: { disable: true } },
};

paginationUrlHref.storyName = "Pagination with URL updates";
paginationUrlHref.parameters = {
  design: {
    type: "figma",
    url:
      "https://www.figma.com/file/qShodlfNCJHb8n03IFyApM/Master?node-id=17226%3A932",
  },
};

const PaginationChangeFunctionTemplate: Story<PaginationProps> = args => {
  const [page, setPage] = useState(args.currentPage);

  const pageUpdate = (page: number) => {
    setPage(page);
    console.log(`Changed. Current page is ${page}`);
  };

  // Allows storybook controls to control currentPage prop.
  useEffect(() => {
    setPage(args.currentPage);
  }, [args.currentPage]);

  return (
    <Pagination
      pageCount={args.pageCount}
      currentPage={page}
      onPageChange={pageUpdate}
    />
  );
};

export const paginationOnClick = PaginationChangeFunctionTemplate.bind({});

paginationOnClick.args = {
  pageCount: 100,
  currentPage: 7,
};

paginationOnClick.argTypes = {
  blockName: { table: { disable: true } },
  className: { table: { disable: true } },
  modifiers: { table: { disable: true } },
};
paginationOnClick.storyName = "Pagination with unchanging URL";
