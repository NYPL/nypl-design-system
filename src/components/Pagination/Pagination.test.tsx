import { expect } from "chai";
import { stub } from "sinon";
import * as Enzyme from "enzyme";
import * as React from "react";
import Pagination from "./Pagination";

describe("Pagination with getPageHref", () => {
  let wrapper: Enzyme.ReactWrapper<any, any>;
  let getPageHref;

  beforeEach(() => {
    getPageHref = (page: number) => {
      return `page=${page}`;
    };
  });

  it("Renders a nav element with items and links", () => {
    wrapper = Enzyme.mount(
      <Pagination pageCount={20} currentPage={6} getPageHref={getPageHref} />
    );
    expect(wrapper.find("nav")).to.have.lengthOf(1);
    expect(wrapper.find("ul")).to.have.lengthOf(1);
    // Previous/Next buttons + truncated item list = 11 total items
    expect(wrapper.find("li")).to.have.lengthOf(11);
    // Each element links to anotherpage
    const links = wrapper.find("a");

    // Pagination should show
    // Previous 1 ... 4 5 6 7 8 ... 20 Next
    expect(links).to.have.lengthOf(9);
    // Previous Page
    expect(links.get(0).props.href).to.equal("page=5");
    expect(links.get(1).props.href).to.equal("page=1");
    expect(links.get(2).props.href).to.equal("page=4");
    expect(links.get(3).props.href).to.equal("page=5");
    expect(links.get(4).props.href).to.equal("page=6");
    expect(links.get(5).props.href).to.equal("page=7");
    expect(links.get(6).props.href).to.equal("page=8");
    expect(links.get(7).props.href).to.equal("page=20");
    // Next Page
    expect(links.get(8).props.href).to.equal("page=7");
  });

  it("Previous link is disabled when on the first page item", () => {
    wrapper = Enzyme.mount(
      <Pagination pageCount={11} currentPage={1} getPageHref={getPageHref} />
    );
    expect(wrapper.find("a").first().hasClass("disabled")).to.equal(true);
    expect(
      wrapper.find("a").first().find("[aria-disabled='true']")
    ).to.have.lengthOf(1);
  });

  it("Next link is disabled when on the last page item", () => {
    wrapper = Enzyme.mount(
      <Pagination pageCount={11} currentPage={11} getPageHref={getPageHref} />
    );
    expect(wrapper.find("a").last().hasClass("disabled")).to.equal(true);
    expect(
      wrapper.find("a").last().find("[aria-disabled='true']")
    ).to.have.lengthOf(1);
  });

  it("Current page item has active class", () => {
    wrapper = Enzyme.mount(
      <Pagination pageCount={11} currentPage={5} getPageHref={getPageHref} />
    );
    expect(wrapper.find("a").at(5).hasClass("selected")).to.equal(true);
  });

  it("When pagination has 1 element, pagination is not shown", () => {
    const shallow = Enzyme.shallow(
      <Pagination pageCount={1} currentPage={1} getPageHref={getPageHref} />
    );
    expect(shallow.isEmptyRender()).to.equal(true);
  });

  it("When pagination has 0 elements, pagination is not shown", () => {
    const shallow = Enzyme.shallow(
      <Pagination pageCount={0} currentPage={1} getPageHref={getPageHref} />
    );

    expect(shallow.isEmptyRender()).to.equal(true);
  });
});

describe("Pagination with changeCallback", () => {
  let wrapper: Enzyme.ReactWrapper<any, any>;
  let changeCallback;

  beforeEach(() => {
    changeCallback = stub();
  });

  it("Renders a nav element with an unordered list of items", () => {
    wrapper = Enzyme.mount(
      <Pagination
        pageCount={11}
        currentPage={6}
        onPageChange={changeCallback}
      />
    );
    expect(wrapper.find("nav")).to.have.lengthOf(1);
    expect(wrapper.find("ul")).to.have.lengthOf(1);
    // Previous/Next buttons + truncated item list = 11 total items
    expect(wrapper.find("li")).to.have.lengthOf(11);
  });

  it("Previous link is disabled when on the first page item", () => {
    wrapper = Enzyme.mount(
      <Pagination
        pageCount={11}
        currentPage={1}
        onPageChange={changeCallback}
      />
    );
    expect(wrapper.find("a").first().hasClass("disabled")).to.equal(true);
    expect(
      wrapper.find("a").first().find("[aria-disabled='true']")
    ).to.have.lengthOf(1);
  });

  it("Next link is disabled when on the last page item", () => {
    wrapper = Enzyme.mount(
      <Pagination
        pageCount={11}
        currentPage={11}
        onPageChange={changeCallback}
      />
    );
    expect(wrapper.find("a").last().hasClass("disabled")).to.equal(true);
    expect(
      wrapper.find("a").last().find("[aria-disabled='true']")
    ).to.have.lengthOf(1);
  });

  it("Current page item has active class", () => {
    wrapper = Enzyme.mount(
      <Pagination
        pageCount={11}
        currentPage={5}
        onPageChange={changeCallback}
      />
    );
    expect(wrapper.find("a").at(5).hasClass("selected")).to.equal(true);
  });

  it("When page item is selected, runs the onPageChange callback", () => {
    wrapper = Enzyme.mount(
      <Pagination
        pageCount={11}
        currentPage={5}
        onPageChange={changeCallback}
      />
    );
    expect(wrapper.find("a").at(2).simulate("click"));
    expect(changeCallback.callCount).to.equal(1);
  });

  it("When pagination has 1 element, pagination is not shown", () => {
    const shallow = Enzyme.shallow(
      <Pagination pageCount={1} currentPage={1} onPageChange={changeCallback} />
    );

    expect(shallow.isEmptyRender()).to.equal(true);
  });

  it("When pagination has 0 elements, pagination is not shown", () => {
    const shallow = Enzyme.shallow(
      <Pagination pageCount={0} currentPage={1} onPageChange={changeCallback} />
    );

    expect(shallow.isEmptyRender()).to.equal(true);
  });
});
