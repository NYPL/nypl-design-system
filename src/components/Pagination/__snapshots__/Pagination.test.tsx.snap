// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Rendering renders the UI snapshot correctly 1`] = `
<nav
  aria-label="Pagination"
  className="css-0"
  id="firstPage"
  role="navigation"
>
  <ul
    className="css-0"
    id="firstPage-list"
  >
    <li>
      <a
        aria-current="page"
        aria-label="Page 1"
        className="css-1858g8z"
        href="page=1"
        id="firstPage-1"
        rel={null}
        target={null}
      >
        1
      </a>
    </li>
    <li>
      <a
        aria-current={null}
        aria-label="Page 2"
        className="css-0"
        href="page=2"
        id="firstPage-2"
        rel={null}
        target={null}
      >
        2
      </a>
    </li>
    <li>
      <a
        aria-current={null}
        aria-label="Page 3"
        className="css-0"
        href="page=3"
        id="firstPage-3"
        rel={null}
        target={null}
      >
        3
      </a>
    </li>
    <li>
      <a
        aria-current={null}
        aria-label="Page 4"
        className="css-0"
        href="page=4"
        id="firstPage-4"
        rel={null}
        target={null}
      >
        4
      </a>
    </li>
    <li>
      <a
        aria-current={null}
        aria-label="Page 5"
        className="css-0"
        href="page=5"
        id="firstPage-5"
        rel={null}
        target={null}
      >
        5
      </a>
    </li>
    <li>
      ...
    </li>
    <li>
      <a
        aria-current={null}
        aria-label="Page 10"
        className="css-0"
        href="page=10"
        id="firstPage-10"
        rel={null}
        target={null}
      >
        10
      </a>
    </li>
    <li>
      <a
        aria-label="Next page"
        className="css-0"
        href="page=2"
        id="firstPage-Next"
        rel={null}
        target={null}
      >
        Next
      </a>
    </li>
  </ul>
</nav>
`;

exports[`Pagination Rendering renders the UI snapshot correctly 2`] = `
<nav
  aria-label="Pagination"
  className="css-0"
  id="lastPage"
  role="navigation"
>
  <ul
    className="css-0"
    id="lastPage-list"
  >
    <li>
      <a
        aria-label="Previous page"
        className="css-0"
        href="page=9"
        id="lastPage-Previous"
        rel={null}
        target={null}
      >
        Previous
      </a>
    </li>
    <li>
      <a
        aria-current={null}
        aria-label="Page 1"
        className="css-0"
        href="page=1"
        id="lastPage-1"
        rel={null}
        target={null}
      >
        1
      </a>
    </li>
    <li>
      ...
    </li>
    <li>
      <a
        aria-current={null}
        aria-label="Page 6"
        className="css-0"
        href="page=6"
        id="lastPage-6"
        rel={null}
        target={null}
      >
        6
      </a>
    </li>
    <li>
      <a
        aria-current={null}
        aria-label="Page 7"
        className="css-0"
        href="page=7"
        id="lastPage-7"
        rel={null}
        target={null}
      >
        7
      </a>
    </li>
    <li>
      <a
        aria-current={null}
        aria-label="Page 8"
        className="css-0"
        href="page=8"
        id="lastPage-8"
        rel={null}
        target={null}
      >
        8
      </a>
    </li>
    <li>
      <a
        aria-current={null}
        aria-label="Page 9"
        className="css-0"
        href="page=9"
        id="lastPage-9"
        rel={null}
        target={null}
      >
        9
      </a>
    </li>
    <li>
      <a
        aria-current="page"
        aria-label="Page 10"
        className="css-1858g8z"
        href="page=10"
        id="lastPage-10"
        rel={null}
        target={null}
      >
        10
      </a>
    </li>
  </ul>
</nav>
`;

exports[`Pagination Rendering renders the UI snapshot correctly 3`] = `
<nav
  aria-label="Pagination"
  className="css-0"
  id="middlePage"
  role="navigation"
>
  <ul
    className="css-0"
    id="middlePage-list"
  >
    <li>
      <a
        aria-label="Previous page"
        className="css-0"
        href="page=4"
        id="middlePage-Previous"
        rel={null}
        target={null}
      >
        Previous
      </a>
    </li>
    <li>
      <a
        aria-current={null}
        aria-label="Page 1"
        className="css-0"
        href="page=1"
        id="middlePage-1"
        rel={null}
        target={null}
      >
        1
      </a>
    </li>
    <li>
      ...
    </li>
    <li>
      <a
        aria-current={null}
        aria-label="Page 4"
        className="css-0"
        href="page=4"
        id="middlePage-4"
        rel={null}
        target={null}
      >
        4
      </a>
    </li>
    <li>
      <a
        aria-current="page"
        aria-label="Page 5"
        className="css-1858g8z"
        href="page=5"
        id="middlePage-5"
        rel={null}
        target={null}
      >
        5
      </a>
    </li>
    <li>
      <a
        aria-current={null}
        aria-label="Page 6"
        className="css-0"
        href="page=6"
        id="middlePage-6"
        rel={null}
        target={null}
      >
        6
      </a>
    </li>
    <li>
      ...
    </li>
    <li>
      <a
        aria-current={null}
        aria-label="Page 10"
        className="css-0"
        href="page=10"
        id="middlePage-10"
        rel={null}
        target={null}
      >
        10
      </a>
    </li>
    <li>
      <a
        aria-label="Next page"
        className="css-0"
        href="page=6"
        id="middlePage-Next"
        rel={null}
        target={null}
      >
        Next
      </a>
    </li>
  </ul>
</nav>
`;
